lxd/container_lxc.go

```diff
@@ -1908,6 +1924,42 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
 	return runConfig, nil
 }
 
+// deviceAttachMounts live attaches mounts to a container.
+func (c *containerLXC) deviceAttachMounts(configCopy map[string]string, mounts []device.MountEntryItem) error {
+	for _, mount := range mounts {
+		flags := 0
+
+		// Convert options into flags.
+		for _, opt := range mount.Opts {
+			if opt == "bind" {
+				flags |= unix.MS_BIND
+			}
+		}
+
+		// Mount it into the container.
+		err := c.insertMount(mount.DevPath, mount.TargetPath, mount.FSType, flags, mount.Shift)
+		if err != nil {
+			return fmt.Errorf("Failed to add mount for device: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// deviceAddCgroupRules live adds cgroup rules to a container.
+func (c *containerLXC) deviceAddCgroupRules(configCopy map[string]string, cgroups []device.RunConfigItem) error {
+	for _, rule := range cgroups {
+		if c.isCurrentlyPrivileged() && !c.state.OS.RunningInUserNS && c.state.OS.CGroupDevicesController {
```

Comment:
Just a nit (not blocker), but this should check that the key is indeed a devices cgroup key and only apply the restriction if that's the case.

We may soon get a gpu cgroup which would also be configured through the devices and would not have the same restriction as the devices croup.

Reply:
@stgraber thats fine, I'll add a prefix match on "devices.". What is the best way to test that these are being set correctly?

Reply:
@stgraber I've used this approach:

1. Run the container as privileged: ```lxc config set c1 security.privileged true```

Then start the container and check the IB devices; major numbers are allowed in the container's cgroup devices.list:

```
root@lantea:/sys/fs/cgroup/devices/lxc.payload/c1# cat devices.list
b *:* m
c *:* m
c 136:* rwm
c 1:3 rwm
c 1:5 rwm
c 1:7 rwm
c 1:8 rwm
c 1:9 rwm
c 5:0 rwm
c 5:1 rwm
c 5:2 rwm
c 10:229 rwm
c 10:200 rwm
c 231:193 rwm
c 231:66 rwm
c 231:2 rwm
root@lantea:/sys/fs/cgroup/devices/lxc.payload/c1# cat /proc/devices 
Character devices:
  1 mem
  4 /dev/vc/0
  4 tty
  4 ttyS
  5 /dev/tty
  5 /dev/console
  5 /dev/ptmx
  5 ttyprintk
  7 vcs
 10 misc
 13 input
 21 sg
 29 fb
 89 i2c
108 ppp
128 ptm
136 pts
180 usb
189 usb_device
204 ttyMAX
226 drm
231 infiniband_mad
231 infiniband_verbs
```